huaweicloud/services/iotda/data_source_huaweicloud_iotda_upgrade_packages.go

```diff
@@ -0,0 +1,180 @@
+			},
+			"packages": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: `The list of the upgrade packages.`,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The ID of the upgrade package.`,
+						},
+						"space_id": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The resource space ID to which the upgrade package belongs.`,
+						},
+						"product_id": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The product ID associated with the upgrade package.`,
+						},
+						"type": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The type of the upgrade package.`,
+						},
+						"description": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The description of the upgrade package.`,
+						},
+						"version": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The version number of the upgrade package.`,
+						},
+						"support_source_versions": {
+							Type:        schema.TypeList,
+							Computed:    true,
+							Elem:        &schema.Schema{Type: schema.TypeString},
+							Description: `The list of source versions that support the upgrade of this version package.`,
+						},
+						"custom_info": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The custom information pushed to the device.`,
+						},
+						"created_at": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The time when the software and firmware packages are uploaded to the IoT platform.`,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceIotdaUpgradePackagesRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	isDerived := WithDerivedAuth(cfg, region)
+	client, err := cfg.HcIoTdaV5Client(region, isDerived)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	var (
+		upgradePackages []model.OtaPackageInfo
+		limit           = int32(50)
+		offset          int32
+	)
+
+	for {
+		listOpts := model.ListOtaPackageInfoRequest{
+			AppId:       utils.StringIgnoreEmpty(d.Get("space_id").(string)),
+			PackageType: d.Get("type").(string),
+			ProductId:   utils.StringIgnoreEmpty(d.Get("product_id").(string)),
+			Version:     utils.StringIgnoreEmpty(d.Get("version").(string)),
+			Limit:       utils.Int32(limit),
+			Offset:      &offset,
+		}
+
+		listResp, listErr := client.ListOtaPackageInfo(&listOpts)
+		if listErr != nil {
+			return diag.Errorf("error retrieving upgrade packages: %s", listErr)
+		}
+
+		if len(*listResp.Packages) == 0 {
```

Comment:
Please make nil judgment before pointer conversion to avoid the risk of null pointer.
```suggestion
                if listResp == nil || listResp.Packages == nil {
                  break
                }
		if len(*listResp.Packages) == 0 {
```